### PR TITLE
DAOS-2061 api: Add daos_cont_set_prop() to client API

### PR DESCRIPTION
**Pull request by kjacque**
The caller must have a container handle with RW capabilities
to set the container properties.

Signed-off-by: Kris Jacque <kristin.jacque@intel.com>

### CODE REVIEW

File: src/container/rpc.h [hunk 1]
```diff
@@ -358,6 +361,15 @@ CRT_RPC_DECLARE(cont_tgt_epoch_aggregate, DAOS_ISEQ_CONT_TGT_EPOCH_AGGREGATE,
 CRT_RPC_DECLARE(cont_tgt_snapshot_notify, DAOS_ISEQ_CONT_TGT_SNAPSHOT_NOTIFY,
 		DAOS_OSEQ_CONT_TGT_SNAPSHOT_NOTIFY)
 
+#define DAOS_ISEQ_CONT_PROP_SET	/* input fields */		 \
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/container/rpc.h]: False alarm

File: src/include/daos_task.h [hunk 2]
```diff
@@ -292,6 +293,11 @@ typedef struct {
 	daos_prop_t		*prop;
 } daos_cont_query_t;
 
+typedef struct {
```
[the remark this answers is not in the document]
Comment (kjacque) [reply src/include/daos_task.h]: Ignoring this one to be consistent with the rest of the structs in this file.

File: src/container/srv_container.c [hunk 3]
```diff
@@ -1468,6 +1468,59 @@ cont_query(struct rdb_tx *tx, struct ds_pool_hdl *pool_hdl, struct cont *cont,
...
+	 * TODO DAOS-2063: Update check when set-prop and set-acl capas added.
+	 */
+	if (!(hdl->ch_capas & DAOS_COO_RW))
+		D_GOTO(out, rc = -DER_NO_PERM);
+
+	rc = cont_prop_write(tx, &cont->c_prop, prop_in);
+	if (rc != 0)
+		D_GOTO(out, rc);
+
+	rc = rdb_tx_commit(tx);
```
Comment (liw) [inline src/container/srv_container.c]: tx is supposed to be committed at most once by cont_op_with_svc. Could we instead read the properties and merge the result with prop_in?

Comment (kjacque) [reply src/container/srv_container.c]: That's a good suggestion, thanks, I'll do that instead. I wasn't sure if the tx commit was wrong here, but it did seem to work on my system.

File: src/container/cli.c [hunk 4]
```diff
@@ -973,6 +973,124 @@ dc_cont_query(tse_task_t *task)
...
+	struct cont_set_prop_args	 arg;
+	int				 rc;
+
+	args = dc_task_get_args(task);
+	D_ASSERTF(args != NULL, "Task Argument OPC does not match DC OPC\n");
+
+	cont = dc_hdl2cont(args->coh);
+	if (cont == NULL)
+		D_GOTO(err, rc = -DER_NO_HDL);
+
```
Comment (kccain) [inline src/container/cli.c]: can cont->dc_capas be checked to see if it includes DAOS_COO_RW in this function, and return an error in this event that the container was not opened with RW permissions?

File: src/tests/suite/daos_container.c [hunk 5]
```diff
@@ -562,6 +562,78 @@ co_acl(void **state)
...
+
+	print_message("Setting the container props\n");
+	rc = daos_cont_set_prop(arg->coh, prop_in, NULL);
+	assert_int_equal(rc, 0);
+
+	print_message("Querying all container props\n");
+	prop_out = daos_prop_alloc(0);
+	assert_non_null(prop_out);
+	rc = daos_cont_query(arg->coh, NULL, prop_out, NULL);
+
```
Comment (kccain) [inline src/tests/suite/daos_container.c]: should the code have assert_int_equal(rc, 0) here?

File: src/container/cli.c (hunk 4 above)
Comment (kjacque) [reply src/container/cli.c]: We could check here, but we must gate access on the server side also, as in the current implementation -- need to cover for the case where a client uses their own modified version of libdaos that takes out the check. Is there a reason to kick out the user early? I'm reluctant to check the same condition twice.

File: src/tests/suite/daos_container.c (hunk 5 above)
Comment (kjacque) [reply src/tests/suite/daos_container.c]: Good catch!

File: src/container/cli.c (hunk 4 above)
Comment (kccain) [reply src/container/cli.c]: Only reason I can think of is to quickly error directly at the client, avoid the RPC trip to the pool service leader server? Not sure it's a big deal. I agree with you the check already in the server side needs to remain there.

I had been looking at some other client-side code I think is must have been cont_open(). it checks if the container open argument flag is RW meanwhile if the pool capas is RO then the client fails directly with -DER_NO_PERM before sending RPC.

Comment (kjacque) [reply src/container/cli.c]: Per our Skype chat, leaving it as-is for now. We can address the client side when we begin enforcing the container ACLs, which will require changes to this check.

File: src/common/prop.c [hunk 6]
```diff
@@ -73,30 +96,77 @@ daos_prop_free(daos_prop_t *prop)
...
-		};
+		daos_prop_entry_free_value(entry);
 	}
 
 	D_FREE(prop->dpp_entries);
 	D_FREE_PTR(prop);
 }
 
+daos_prop_t *
+daos_prop_merge(daos_prop_t* old_prop, daos_prop_t *new_prop)
```
[the remark this answers is not in the document]
Comment (liw) [reply src/common/prop.c]: daosbuild1 is right this time. :) No need to refresh just for this though.

File: src/include/daos_prop.h [hunk 7]
```diff
@@ -273,6 +273,17 @@ daos_prop_alloc(uint32_t entries_nr);
...
+/**
+ * Merge a set of new DAOS properties into a set of existing DAOS properties.
+ *
+ * \param[in]	old_prop	Existing set of properties
+ * \param[in]	new_prop	New properties - may override old entries
+ *
+ * \return	Newly allocated merged property
+ */
+daos_prop_t *
+daos_prop_merge(daos_prop_t* old_prop, daos_prop_t *new_prop);
```
[the remark this answers is not in the document]
Comment (liw) [reply src/include/daos_prop.h]: Ditto.

File: src/common/prop.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (kjacque) [reply src/common/prop.c]: I did fix this, just didn't note it when I did it.

File: src/include/daos_prop.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (kjacque) [reply src/include/daos_prop.h]: Fixed this one, too.

**Comment by kjacque** [comment]
The Functional test failure shows another instance of DAOS-3945.

**Pull request merged**